python/pyspark/sql/dataframe.py

```diff
@@ -2153,6 +2153,50 @@ def transform(self, func):
                                               "should have been DataFrame." % type(result)
         return result
 
+    @since(3.1)
+    def sameSemantics(self, other):
+        """
+        Return true when the query plan of the given :class:`DataFrame` will return the same
+        results as this :class:`DataFrame`.
+
+        >>> df1 = spark.createDataFrame([(1, 2),(4, 5)], ["col1", "col2"])
+        >>> df2 = spark.createDataFrame([(1, 2),(4, 5)], ["col0", "col2"])
+        >>> df3 = spark.createDataFrame([(0, 2),(4, 5)], ["col1", "col2"])
+        >>> df4 = spark.createDataFrame([(1, 2),(4, 5)], ["col1", "col2"])
+        >>> df1.sameSemantics(df2)
+        False
+        >>> df1.sameSemantics(df3)
+        False
+        >>> df1.sameSemantics(df4)
+        True
+        >>> df1.sameSemantics(df1)
+        True
+        """
+        if not isinstance(other, DataFrame):
+            raise ValueError("other parameter should be of DataFrame; however, got %s"
+                             % type(other))
+        return self._jdf.sameSemantics(other._jdf)
+
+    @since(3.1)
+    def semanticHash(self):
+        """
+        Returns a `hashCode` for the calculation performed by the query plan of this Dataset.
+
+        >>> df1 = spark.createDataFrame([(1, 2),(4, 5)], ["col1", "col2"])
+        >>> df2 = spark.createDataFrame([(1, 2),(4, 5)], ["col0", "col2"])
+        >>> df3 = spark.createDataFrame([(0, 2),(4, 5)], ["col1", "col2"])
+        >>> df4 = spark.createDataFrame([(1, 2),(4, 5)], ["col1", "col2"])
+        >>> df1.semanticHash() == df2.semanticHash()
+        False
+        >>> df1.semanticHash() == df3.semanticHash()
+        False
+        >>> df1.semanticHash() == df4.semanticHash()
+        True
```

Comment:
Currently, this test would fail. I don't have a clue why only the same object would have the same hash and all other cases seem to have a different hash. Scala side tests all pass.

Reply:
Do you have any ideas? @HyukjinKwon

Reply:
I need to debug too. Don't know the cause. One hypothesis on my mind is though,

There is one difference between Python and Scala side is, the Python side is currently creating the `DataFrame` based on `RDD[Array[Byte]]` in JVM perspective. 
In Scala side, it will hold `LocalRelation` which contains the actual data. This could affect the hash.

Reply:
If this is the case, we might have to leave an explicit caveat that false positive case possible. Also we might have to mark those APIs as `@Unstable` or `@Experimental` due to this reason cc @mengxr.

Reply:
emm, let's change unit test. Don't test on dataframe created from in-memory list (LocalRelation), they have different implementation between scala and pyspark.
Our usecase also do not care the behavior of LocalRelation.
so, I suggested add a unit test on: spark.df.read(...).where(...)